src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala

```diff
@@ -69,19 +69,23 @@ object PV63 {
             RLPList(RLPValue(hpEncode(keyNibbles.toArray[Byte], isLeaf = false)), child.fold(_.hash, _.value): RLPEncodeable)
           case n: MptBranch =>
             import n._
-            RLPList(children.map(e => e.fold(_.hash, _.value)).map(e => byteStringEncDec.encode(e)) :+ (terminator: RLPEncodeable): _*)
+            RLPList(children.map(e => e.fold(a => byteStringEncDec.encode(a.hash), b => rawDecode(b.value.toArray[Byte]))) :+ (terminator: RLPEncodeable): _*)
         }
       }
 
       override def decode(rlp: RLPEncodeable): MptNode = rlp match {
         case rlpList: RLPList if rlpList.items.length == BranchNodeChildLength + 1 =>
-          MptBranch(rlpList.items.take(BranchNodeChildLength).map { bytes =>
-            val v = rlpDecode[ByteString](bytes)
-            if (v.nonEmpty && v.length < 32) {
-              Right(MptValue(v))
-            } else {
-              Left(MptHash(v))
-            }
+          MptBranch(rlpList.items.take(BranchNodeChildLength).map {
+            case bytes: RLPValue =>
+              val v = rlpDecode[ByteString](bytes)
+              if (v.nonEmpty && v.length < 32) {
+                Right(MptValue(ByteString(rlpEncode(bytes))))
```

Comment:
I'm not sure about that, regarding the extension nodes having values, shouldn't it have a `MptValue` for the case where the `child` is a branch node that's small enough? As far as I know if that branch node, encoded, is smaller than 32 bytes it should be stored encoded in the extension node.

I searched for that in geth (`trie/triedbmut.rs`) and branch node children and extension node child have the same format:
```
enum Node {
	...
	Extension(NodeKey, NodeHandle),
	Branch(Box<[Option<NodeHandle>; 16]>, Option<DBValue>)	
}
```